include/deal.II/base/cuda_tensor.cuh

```diff
@@ -0,0 +1,340 @@
+   * rank-2 tensor is a matrix with dim-by-dim components, etc. There are
+   * specializations of this class for rank-0 and rank-1 tensors. There is also
+   * a related class SymmetricTensor for tensors of even rank whose elements are
+   * symmetric.
+   *
+   * @tparam dim An integer that denotes the dimension of the space in which
+   * this tensor operates. This of course equals the number of coordinates that
+   * identify a point and rank-1 tensor.
+   *
+   * @tparam Number The data type in which the tensor elements are to be stored.
+   * This will, in almost all cases, simply be the default @p double, but there
+   * are cases where one may want to store elements in a different (and always
+   * scalar) type. It can be used to base tensors on @p float or @p complex
+   * numbers or any other data type that implements basic arithmetic operations.
+   * Another example would be a type that allows for Automatic Differentiation
+   * (see, for example, the Sacado type used in step-33) and thereby can
+   * generate analytic (spatial) derivatives of a function that takes a tensor
+   * as argument.
+   *
+   * @ingroup CUDAWrappers
+   *
+   * @author Karl Ljungkvist, Bruno Turcksin, 2016
+   */
+  template <int rank_, int dim, typename Number>
+  class Tensor
+  {
+  public:
+    /**
+     * Type of objects encapsulated by this container and returned by
+     * operator[](). This is a tensor of lower rank for a general tensor, and
+     * a scalar number type of Tensor<1,dim,Number>.
+     */
+    typedef typename Tensor<rank_-1,dim,Number>::tensor_type value_type;
+
+    /**
+     * Constructor. Initialize all entries to zero.
+     */
+    __host__ __device__ Tensor() = default;
+
+    /**
+     * Read-Write access operator.
+     */
+    __host__ __device__ Tensor<rank_,dim,Number>::value_type &operator[](const unsigned int i);
+
+    /**
+     * Read-only access operator.
+     */
+    __host__ __device__ const Tensor<rank_,dim,Number>::value_type &operator[](const unsigned int i) const;
+
+    /**
+     * Return the square of the Frobenius-norm of a tensor, i.e., the sum of
+     * the absolute squares of all entries.
+     */
+    __host__ __device__ Number norm_square() const;
+
+    /**
+     * Scale the tensor by @p factor, i.e., multiplu all components by @p
+     * factor.
+     */
+    template <typename Number2>
+    __host__ __device__ Tensor<rank_,dim,Number> &operator*=(const Number2 factor);
+
+    /**
+     * Internal type declaration that is used to specialize the return type of
+     * operator[]() for Tensor<1,dim,Number>
+     */
+    typedef Tensor<rank_, dim, Number> tensor_type;
+
+  private:
+    /**
+     * Array of tensors holding the subelements.
+     */
+    Tensor<rank_-1, dim, Number> values[(dim != 0) ? dim : 1];
+    // ... avoid a compiler warning in case of dim == 0 and ensure that the
+    // array always has positive size.
+  };
```

Comment:
Yes, it might possible using a macro. Something like `DEAL_II_HOST_DEV` which would expand to `__host__ __device__` if cuda is enable and would be empty otherwise.